File: scripts/zccache-contract.js
#!/usr/bin/env node
"use strict";

const crypto = require("crypto");
const fs = require("fs");
const path = require("path");

const PACKAGE_ROOT = path.resolve(__dirname, "..");
const CONTRACT_PATH = path.join(PACKAGE_ROOT, "contracts", "zccache-runtime.v1.json");
const CONTRACT = JSON.parse(fs.readFileSync(CONTRACT_PATH, "utf8"));

const ARCHIVE_EXT = CONTRACT.release_archive.extension;
const MANIFEST_NAME = CONTRACT.release_archive.manifest_name;
const MANIFEST_MIN_SCHEMA_VERSION = CONTRACT.release_archive.manifest_min_schema_version;
const RELEASE_BUNDLED_BINARIES = Object.freeze([...CONTRACT.release_archive.required_binaries]);
// zccache is compiled into soldr (soldr#1368); no standalone zccache
// binaries are bundled or downloaded, so this list is empty.
const ZCCACHE_BUNDLED_BINARIES = Object.freeze([
  ...(CONTRACT.zccache.required_binaries || []),
]);
const CRGX_BUNDLED_BINARY = CONTRACT.crgx.required_binaries[0];
const CARGO_CHEF_BUNDLED_BINARY = CONTRACT.cargo_chef.required_binaries[0];

function binaryName(baseName, platform = process.platform) {
  return `${baseName}${platform === "win32" ? ".exe" : ""}`;
}

function releaseBinaryNames(platform = process.platform) {
  return RELEASE_BUNDLED_BINARIES.map((baseName) => binaryName(baseName, platform));
}

function zccacheTargetForSoldrTarget(soldrTarget) {
  if (!soldrTarget.includes("-unknown-linux-")) {
    return soldrTarget;
  }
  const arch = soldrTarget.split("-unknown-linux-", 1)[0];
  return `${arch}-unknown-linux-${CONTRACT.release_archive.linux_zccache_target_libc}`;
}

function sha256File(filePath) {
  return crypto.createHash("sha256").update(fs.readFileSync(filePath)).digest("hex");
}

function collectManifestBinaries(manifest) {
  const binaries = new Map();
  if (manifest.soldr && manifest.soldr.binary && manifest.soldr.sha256) {
    binaries.set(manifest.soldr.binary, manifest.soldr.sha256);
  }
  const zccacheBinaries = (manifest.zccache && manifest.zccache.binaries) || [];
  for (const entry of zccacheBinaries) {
    if (entry && entry.name && entry.sha256) {
      binaries.set(entry.name, entry.sha256);
    }
  }
  if (manifest.crgx && manifest.crgx.binary && manifest.crgx.sha256) {
    binaries.set(manifest.crgx.binary, manifest.crgx.sha256);
  }
  if (manifest.cargo_chef && manifest.cargo_chef.binary && manifest.cargo_chef.sha256) {
    binaries.set(manifest.cargo_chef.binary, manifest.cargo_chef.sha256);
  }
  return binaries;
}

function soldrDebugInfoEntries(manifest) {
  const entries = manifest.soldr && Array.isArray(manifest.soldr.debug_info)
    ? manifest.soldr.debug_info
    : [];
  return entries.filter((entry) => entry && entry.name && entry.sha256);
}

function validateReleaseManifest(manifest, options) {
  const { soldrTarget, platform = process.platform, findFile } = options;
  if (!Number.isInteger(manifest.schema_version) || manifest.schema_version < MANIFEST_MIN_SCHEMA_VERSION) {
    throw new Error(`release manifest schema_version must be >= ${MANIFEST_MIN_SCHEMA_VERSION}`);
  }
  if (!manifest.archive || manifest.archive.format !== ARCHIVE_EXT) {
    throw new Error(`release manifest archive.format must be ${ARCHIVE_EXT}`);
  }
  if (!manifest.soldr || manifest.soldr.target !== soldrTarget) {
    throw new Error(`release manifest soldr.target must be ${soldrTarget}`);
  }
  // zccache is compiled into soldr (soldr#1368) — no standalone zccache
  // block is required in the release manifest anymore.
  if (!manifest.crgx || manifest.crgx.target !== soldrTarget) {
    throw new Error(`release manifest crgx.target must be ${soldrTarget}`);
  }
  if (!manifest.cargo_chef || manifest.cargo_chef.target !== soldrTarget) {
    throw new Error(`release manifest cargo_chef.target must be ${soldrTarget}`);
  }

  const expectedNames = releaseBinaryNames(platform);
  const manifestBinaries = collectManifestBinaries(manifest);
  for (const name of expectedNames) {
    if (!manifestBinaries.has(name)) {
      throw new Error(`release manifest is missing bundled binary record for ${name}`);
    }
    const expectedSha = String(manifestBinaries.get(name)).toLowerCase();
    if (!/^[0-9a-f]{64}$/.test(expectedSha)) {
      throw new Error(`release manifest sha256 for ${name} is not lowercase hex`);
    }
    const filePath = findFile(name);
    const actualSha = sha256File(filePath);
    if (actualSha !== expectedSha) {
      throw new Error(`release manifest sha256 mismatch for ${name}: expected ${expectedSha}, got ${actualSha}`);
    }
  }

  const debugInfo = soldrDebugInfoEntries(manifest);
  if (platform === "win32" && debugInfo.length === 0) {
    throw new Error("release manifest is missing soldr debug_info PDB entry");
  }
  for (const entry of debugInfo) {
    if (entry.format !== "pdb") {
      throw new Error(`unsupported soldr debug_info format for ${entry.name}: ${entry.format}`);
    }
    if (!/\.pdb$/i.test(entry.name)) {
      throw new Error(`soldr debug_info entry must name a .pdb file: ${entry.name}`);
    }
    const expectedSha = String(entry.sha256).toLowerCase();
    if (!/^[0-9a-f]{64}$/.test(expectedSha)) {
      throw new Error(`release manifest sha256 for ${entry.name} is not lowercase hex`);
    }
    const filePath = findFile(entry.name);
    const actualSha = sha256File(filePath);
    if (actualSha !== expectedSha) {
      throw new Error(`release manifest sha256 mismatch for ${entry.name}: expected ${expectedSha}, got ${actualSha}`);
    }
  }
}

module.exports = {
  ARCHIVE_EXT,
  CONTRACT,
  CONTRACT_PATH,
  CARGO_CHEF_BUNDLED_BINARY,
  CRGX_BUNDLED_BINARY,
  MANIFEST_NAME,
  RELEASE_BUNDLED_BINARIES,
  ZCCACHE_BUNDLED_BINARIES,
  binaryName,
  releaseBinaryNames,
  soldrDebugInfoEntries,
  validateReleaseManifest,
  zccacheTargetForSoldrTarget,
};

File: .github/actions/setup-soldr/zccache_contract.py
#!/usr/bin/env python3
from __future__ import annotations

import hashlib
import json
from pathlib import Path
from typing import Any

REPO_ROOT = Path(__file__).resolve().parents[3]
CONTRACT_PATH = REPO_ROOT / "contracts" / "zccache-runtime.v1.json"


def load_contract(path: Path = CONTRACT_PATH) -> dict[str, Any]:
    return json.loads(path.read_text(encoding="utf-8"))


CONTRACT = load_contract()
ARCHIVE_EXT = str(CONTRACT["release_archive"]["extension"])
MANIFEST_NAME = str(CONTRACT["release_archive"]["manifest_name"])
MANIFEST_MIN_SCHEMA_VERSION = int(
    CONTRACT["release_archive"]["manifest_min_schema_version"]
)
ZCCACHE_BUNDLED_BINARIES = tuple(
    CONTRACT.get("zccache", {}).get("required_binaries", ())
)
CRGX_BUNDLED_BINARY = str(CONTRACT["crgx"]["required_binaries"][0])
CARGO_CHEF_BUNDLED_BINARY = str(CONTRACT["cargo_chef"]["required_binaries"][0])
RELEASE_BUNDLED_BINARIES = tuple(CONTRACT["release_archive"]["required_binaries"])
CRGX_LOCAL_DIR_ENV = str(CONTRACT["crgx"]["local_dir_env"])
CARGO_CHEF_LOCAL_DIR_ENV = str(CONTRACT["cargo_chef"]["local_dir_env"])


def binary_name(base: str, *, windows: bool) -> str:
    return f"{base}.exe" if windows else base


def release_binary_names(*, windows: bool) -> tuple[str, ...]:
    return tuple(
        binary_name(base, windows=windows) for base in RELEASE_BUNDLED_BINARIES
    )


def zccache_target_for_soldr_target(soldr_target: str) -> str:
    if bool(CONTRACT.get("zccache", {}).get("embedded")):
        return soldr_target
    if "-unknown-linux-" not in soldr_target:
        return soldr_target
    arch, _, _ = soldr_target.partition("-unknown-linux-")
    libc = CONTRACT["release_archive"]["linux_zccache_target_libc"]
    return f"{arch}-unknown-linux-{libc}"


def _sha256_file(path: Path) -> str:
    hasher = hashlib.sha256()
    with path.open("rb") as fh:
        for chunk in iter(lambda: fh.read(1024 * 1024), b""):
            hasher.update(chunk)
    return hasher.hexdigest()


def _manifest_binaries(manifest: dict[str, Any]) -> dict[str, str]:
    binaries: dict[str, str] = {}
    soldr = manifest.get("soldr", {})
    if isinstance(soldr, dict):
        binary = soldr.get("binary")
        sha = soldr.get("sha256")
        if isinstance(binary, str) and isinstance(sha, str):
            binaries[binary] = sha
        sidecars = soldr.get("sidecars", [])
        if isinstance(sidecars, list):
            for entry in sidecars:
                if not isinstance(entry, dict):
                    continue
                name = entry.get("name")
                sha = entry.get("sha256")
                if isinstance(name, str) and isinstance(sha, str):
                    binaries[name] = sha
    zccache = manifest.get("zccache", {})
    if isinstance(zccache, dict):
        for entry in zccache.get("binaries", []):
            if not isinstance(entry, dict):
                continue
            name = entry.get("name")
            sha = entry.get("sha256")
            if isinstance(name, str) and isinstance(sha, str):
                binaries[name] = sha
    crgx = manifest.get("crgx", {})
    if isinstance(crgx, dict):
        binary = crgx.get("binary")
        sha = crgx.get("sha256")
        if isinstance(binary, str) and isinstance(sha, str):
            binaries[binary] = sha
    cargo_chef = manifest.get("cargo_chef", {})
    if isinstance(cargo_chef, dict):
        binary = cargo_chef.get("binary")
        sha = cargo_chef.get("sha256")
        if isinstance(binary, str) and isinstance(sha, str):
            binaries[binary] = sha
    return binaries


def soldr_debug_info_entries(manifest: dict[str, Any]) -> list[dict[str, Any]]:
    soldr = manifest.get("soldr", {})
    if not isinstance(soldr, dict):
        return []
    entries = soldr.get("debug_info", [])
    if not isinstance(entries, list):
        return []
    return [
        entry
        for entry in entries
        if isinstance(entry, dict)
        and isinstance(entry.get("name"), str)
        and isinstance(entry.get("sha256"), str)
    ]


def validate_release_manifest(
    manifest: dict[str, Any],
    *,
    soldr_target: str,
    windows: bool,
    extract_dir: Path,
) -> None:
    schema_version = manifest.get("schema_version")
    if (
        not isinstance(schema_version, int)
        or schema_version < MANIFEST_MIN_SCHEMA_VERSION
    ):
        raise RuntimeError(
            f"release manifest schema_version must be >= {MANIFEST_MIN_SCHEMA_VERSION}"
        )
    archive = manifest.get("archive")
    if not isinstance(archive, dict) or archive.get("format") != ARCHIVE_EXT:
        raise RuntimeError(f"release manifest archive.format must be {ARCHIVE_EXT}")
    soldr = manifest.get("soldr")
    if not isinstance(soldr, dict) or soldr.get("target") != soldr_target:
        raise RuntimeError(f"release manifest soldr.target must be {soldr_target}")
    zccache = manifest.get("zccache")
    expected_zccache_target = zccache_target_for_soldr_target(soldr_target)
    if (
        not isinstance(zccache, dict)
        or zccache.get("target") != expected_zccache_target
    ):
        raise RuntimeError(
            f"release manifest zccache.target must be {expected_zccache_target}"
        )
    crgx = manifest.get("crgx")
    if not isinstance(crgx, dict) or crgx.get("target") != soldr_target:
        raise RuntimeError(f"release manifest crgx.target must be {soldr_target}")
    cargo_chef = manifest.get("cargo_chef")
    if not isinstance(cargo_chef, dict) or cargo_chef.get("target") != soldr_target:
        raise RuntimeError(f"release manifest cargo_chef.target must be {soldr_target}")

    expected_names = set(release_binary_names(windows=windows))
    manifest_binaries = _manifest_binaries(manifest)
    missing = sorted(expected_names.difference(manifest_binaries))
    if missing:
        raise RuntimeError(
            f"release manifest is missing bundled binary records: {', '.join(missing)}"
        )

    for name in sorted(expected_names):
        expected_sha = manifest_binaries[name].lower()
        if len(expected_sha) != 64 or any(
            ch not in "0123456789abcdef" for ch in expected_sha
        ):
            raise RuntimeError(
                f"release manifest sha256 for {name} is not lowercase hex"
            )
        binary_path = locate_extracted_file(extract_dir, name)
        actual_sha = _sha256_file(binary_path)
        if actual_sha != expected_sha:
            raise RuntimeError(
                f"release manifest sha256 mismatch for {name}: expected {expected_sha}, got {actual_sha}"
            )

    debug_info = soldr_debug_info_entries(manifest)
    if windows and not debug_info:
        raise RuntimeError("release manifest is missing soldr debug_info PDB entry")
    for entry in debug_info:
        name = str(entry["name"])
        if entry.get("format") != "pdb":
            raise RuntimeError(
                f"unsupported soldr debug_info format for {name}: {entry.get('format')}"
            )
        if not name.lower().endswith(".pdb"):
            raise RuntimeError(f"soldr debug_info entry must name a .pdb file: {name}")
        expected_sha = str(entry["sha256"]).lower()
        if len(expected_sha) != 64 or any(
            ch not in "0123456789abcdef" for ch in expected_sha
        ):
            raise RuntimeError(
                f"release manifest sha256 for {name} is not lowercase hex"
            )
        actual_sha = _sha256_file(locate_extracted_file(extract_dir, name))
        if actual_sha != expected_sha:
            raise RuntimeError(
                f"release manifest sha256 mismatch for {name}: expected {expected_sha}, got {actual_sha}"
            )


def locate_extracted_file(root: Path, file_name: str) -> Path:
    for candidate in root.rglob(file_name):
        if candidate.is_file():
            return candidate
    raise RuntimeError(f"downloaded archive did not contain {file_name}")
